File: flutter_app/lib/models/worker.dart
import 'dart:math' as math;
import 'package:flutter/material.dart';
import '../core/design_system/app_colors.dart';
import 'worker_dto.dart';

class Worker {
  const Worker({
    required this.id,
    required this.initials,
    required this.name,
    required this.trade,
    required this.phone,
    this.bio,
    required this.location,
    required this.distanceKm,
    required this.vouched,
    required this.vouchScore,
    required this.jobs,
    required this.rating,
    required this.tags,
    required this.verifiedTagIndices,
    required this.emoji,
    required this.gradientStartHex,
    required this.gradientEndHex,
    required this.isVerified,
    this.avatarURL,
    this.reelPlaybackURL,
    this.reelThumbnailURL,
    this.reelDurationSeconds,
  });

  final String id;
  final String initials;
  final String name;
  final String trade;
  final String phone;
  final String? bio;
  final String location;
  final double distanceKm;
  final int vouched;
  final int vouchScore;
  final int jobs;
  final String rating;
  final List<String> tags;
  final Set<int> verifiedTagIndices;
  final String emoji;
  final String gradientStartHex;
  final String gradientEndHex;
  final bool isVerified;
  final String? avatarURL;
  final String? reelPlaybackURL;
  final String? reelThumbnailURL;
  final double? reelDurationSeconds;

  String get category => trade;
  Color get gradientStart => AppColors.fromHex(gradientStartHex);
  Color get gradientEnd => AppColors.fromHex(gradientEndHex);

  static const radiusSteps = [5, 10, 15, 20, 30, 40, 50, 60, 75, 100];

  static const allCategories = [
    'All', 'Electrician', 'Plumber', 'Carpenter', 'Painter', 'Mason', 'Welder',
    'Mechanic', 'Driver', 'Cook', 'Nurse', 'Tailor', 'AC Technician',
    'Interior Decorator', 'Appliance Repair', 'Gardener', 'Security Guard', 'Other',
  ];

  factory Worker.fromDto(WorkerDto dto, {required double userLat, required double userLng}) {
    final style = TradeStyle.forTrade(dto.trade);
    final parts = dto.name.trim().split(RegExp(r'\s+'));
    final initials = parts.map((w) => w.isNotEmpty ? w[0] : '').take(2).join().toUpperCase();
    final distance = (dto.lat != null && dto.lng != null)
        ? haversineKm(userLat, userLng, dto.lat!, dto.lng!)
        : 0.0;
    final cityLabel = dto.city ?? 'Mumbai';
    final location = (dto.lat != null && dto.lng != null)
        ? '$cityLabel · ${distance.toStringAsFixed(1)} km'
        : cityLabel;
    final derivedJobs = math.max(12, dto.vouchScore * 6);
    final r = math.min(5.0, 4.0 + (dto.vouchScore / 100.0));

    return Worker(
      id: dto.id,
      initials: initials,
      name: dto.name,
      trade: dto.trade,
      phone: dto.phone,
      bio: dto.bio,
      location: location,
      distanceKm: distance,
      vouched: dto.vouchScore,
      vouchScore: dto.vouchScore,
      jobs: derivedJobs,
      rating: r.toStringAsFixed(1),
      tags: dto.skillTags,
      verifiedTagIndices: dto.verifiedTagIndices.toSet(),
      emoji: style.emoji,
      gradientStartHex: style.startHex,
      gradientEndHex: style.endHex,
      isVerified: dto.isVerified ?? false,
      avatarURL: dto.avatarURL,
      reelPlaybackURL: dto.reelHlsURL ?? dto.reelURL,
      reelThumbnailURL: dto.reelThumbnailURL,
      reelDurationSeconds: dto.reelDurationSeconds,
    );
  }
}

class TradeStyle {
  const TradeStyle(this.emoji, this.startHex, this.endHex);
  final String emoji;
  final String startHex;
  final String endHex;

  static TradeStyle forTrade(String trade) {
    final lower = trade.toLowerCase();
    if (lower.contains('electric')) return const TradeStyle('⚡', '#0c1829', '#1a3355');
    if (lower.contains('plumb')) return const TradeStyle('🔧', '#081420', '#102840');
    if (lower.contains('hvac') || lower.contains('ac ')) return const TradeStyle('❄️', '#061a0e', '#0c3018');
    if (lower.contains('interior') || lower.contains('decorator')) return const TradeStyle('🎨', '#1e0f05', '#3d2210');
    if (lower.contains('weld')) return const TradeStyle('🔥', '#1a0800', '#361400');
    if (lower.contains('nurse')) return const TradeStyle('🏥', '#0e0718', '#1c1030');
    if (lower.contains('carpenter')) return const TradeStyle('🪚', '#1a0f00', '#33200a');
    if (lower.contains('paint')) return const TradeStyle('🎨', '#0a0a1e', '#16163a');
    if (lower.contains('mason')) return const TradeStyle('🧱', '#1f1208', '#3a2412');
    if (lower.contains('cook')) return const TradeStyle('🍳', '#1c0a00', '#3a1c08');
    if (lower.contains('driver')) return const TradeStyle('🚗', '#0a0a0a', '#1c1c1c');
    if (lower.contains('guard') || lower.contains('security')) return const TradeStyle('🛡️', '#0a0a14', '#1c1c30');
    if (lower.contains('tailor')) return const TradeStyle('🧵', '#160a14', '#2e1830');
    if (lower.contains('garden')) return const TradeStyle('🌿', '#0a1c0a', '#163018');
    if (lower.contains('appliance') || lower.contains('repair')) return const TradeStyle('🔌', '#0a141a', '#162a36');
    if (lower.contains('mechanic')) return const TradeStyle('🔩', '#0a0a0a', '#262626');
    return const TradeStyle('🛠️', '#161616', '#2d2d2d');
  }
}

double haversineKm(double lat1, double lng1, double lat2, double lng2) {
  const r = 6371.0;
  final phi1 = lat1 * math.pi / 180;
  final phi2 = lat2 * math.pi / 180;
  final dphi = (lat2 - lat1) * math.pi / 180;
  final dlam = (lng2 - lng1) * math.pi / 180;
  final a = math.sin(dphi / 2) * math.sin(dphi / 2) +
      math.cos(phi1) * math.cos(phi2) * math.sin(dlam / 2) * math.sin(dlam / 2);
  return r * 2 * math.atan2(math.sqrt(a), math.sqrt(1 - a));
}

File: src/models/worker.js
import { APIConfig } from '../api/config';

const GRADIENTS = [
  '#ee6c4d', '#118AB2', '#C9184A', '#C1440E', '#7B2D8B',
  '#2D6A4F', '#E85D04', '#5C4D7D',
];

/** Mirrors iOS WorkerDTO TradeStyle.forTrade — explore card visuals. */
function tradeStyle(trade = '') {
  const lower = trade.toLowerCase();
  if (lower.includes('electric')) return { emoji: '⚡', gradientStartHex: '#0c1829' };
  if (lower.includes('plumb')) return { emoji: '🔧', gradientStartHex: '#081420' };
  if (lower.includes('hvac') || lower.includes('ac ')) return { emoji: '❄️', gradientStartHex: '#061a0e' };
  if (lower.includes('interior') || lower.includes('decorator')) return { emoji: '🎨', gradientStartHex: '#1e0f05' };
  if (lower.includes('weld')) return { emoji: '🔥', gradientStartHex: '#1a0800' };
  if (lower.includes('nurse')) return { emoji: '🏥', gradientStartHex: '#0e0718' };
  if (lower.includes('carpenter')) return { emoji: '🪚', gradientStartHex: '#1a0f00' };
  if (lower.includes('paint')) return { emoji: '🎨', gradientStartHex: '#0a0a1e' };
  if (lower.includes('mason')) return { emoji: '🧱', gradientStartHex: '#1f1208' };
  if (lower.includes('cook')) return { emoji: '🍳', gradientStartHex: '#1c0a00' };
  if (lower.includes('driver')) return { emoji: '🚗', gradientStartHex: '#0a0a0a' };
  if (lower.includes('guard') || lower.includes('security')) return { emoji: '🛡️', gradientStartHex: '#0a0a14' };
  if (lower.includes('tailor')) return { emoji: '🧵', gradientStartHex: '#160a14' };
  if (lower.includes('garden')) return { emoji: '🌿', gradientStartHex: '#0a1c0a' };
  if (lower.includes('appliance') || lower.includes('repair')) return { emoji: '🔌', gradientStartHex: '#0a141a' };
  if (lower.includes('mechanic')) return { emoji: '🔩', gradientStartHex: '#0a0a0a' };
  return { emoji: '🛠️', gradientStartHex: '#161616' };
}

function hashId(id) {
  let h = 0;
  for (let i = 0; i < (id || '').length; i++) h = (h * 31 + id.charCodeAt(i)) | 0;
  return Math.abs(h);
}

function haversineKm(lat1, lng1, lat2, lng2) {
  if (lat1 == null || lng1 == null || lat2 == null || lng2 == null) return null;
  const R = 6371;
  const dLat = ((lat2 - lat1) * Math.PI) / 180;
  const dLng = ((lng2 - lng1) * Math.PI) / 180;
  const a =
    Math.sin(dLat / 2) ** 2 +
    Math.cos((lat1 * Math.PI) / 180) *
      Math.cos((lat2 * Math.PI) / 180) *
      Math.sin(dLng / 2) ** 2;
  return R * 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
}

function initials(name) {
  return (name || '?')
    .split(/\s+/)
    .filter(Boolean)
    .slice(0, 2)
    .map(w => w[0]?.toUpperCase() ?? '')
    .join('');
}

/**
 * Map API WorkerDTO → UI worker object (mirrors iOS Worker(dto:)).
 */
export function workerFromDTO(dto, userLat, userLng) {
  const dist = haversineKm(userLat, userLng, dto.lat, dto.lng);
  const vouch = dto.vouch_score ?? 0;
  const years = dto.extracted_years;
  const tradeLine = years
    ? `${dto.trade} · ${years} yrs exp`
    : dto.trade;
  const style = tradeStyle(dto.trade || dto.extracted_trade || '');

  return {
    id: dto.id,
    name: dto.name,
    phone: dto.phone,
    trade: tradeLine,
    tradeRaw: dto.trade,
    bio: dto.bio,
    location: dto.city || '—',
    city: dto.city,
    lat: dto.lat,
    lng: dto.lng,
    rating: (vouch / 20).toFixed(1),
    jobs: Math.max(1, Math.round(vouch / 2)),
    vouchScore: vouch,
    vouched: vouch,
    tags: dto.skill_tags || [],
    verifiedTagIndices: new Set(dto.verified_tag_indices || []),
    emoji: style.emoji,
    gradientStartHex: style.gradientStartHex,
    gradientEndHex: GRADIENTS[hashId(dto.id) % GRADIENTS.length],
    initials: initials(dto.name),
    distance: dist != null ? dist.toFixed(1) : '—',
    avatarUrl: dto.avatar_url,
    reelUrl: dto.reel_url,
    reelHlsUrl: dto.reel_hls_url,
    playbackUrl: dto.playback_url,
    isVerified: dto.is_verified ?? false,
    analysisStatus: dto.analysis_status,
    extractedTrade: dto.extracted_trade,
    extractedYears: dto.extracted_years,
    extractedSkills: dto.extracted_skills || [],
    extractedBio: dto.extracted_bio,
    reelVerification: dto.reel_verification,
    isActive: dto.is_active,
    _dto: dto,
  };
}

export function workerUpdatePayload(fields) {
  const p = {};
  if (fields.name != null) p.name = fields.name;
  if (fields.trade != null) p.trade = fields.trade;
  if (fields.bio != null) p.bio = fields.bio;
  if (fields.skillTags != null) p.skill_tags = fields.skillTags;
  if (fields.city != null) p.city = fields.city;
  if (fields.lat != null) p.lat = fields.lat;
  if (fields.lng != null) p.lng = fields.lng;
  return p;
}

export async function withLocation(mapper) {
  const { locationService } = await import('../services/locationService');
  const { lat, lng } = await locationService.current();
  return mapper(lat, lng);
}
